### PR TITLE
refactor: initializers module

### LINKED ISSUE #81: REFACTOR: Initializers module
The current version of the module is not suitable for several reasons:
- Duplicate code for various scoring functions (AIC, BIC, Likelihood)
- A full enumeration does not take into account components of the same family.
- The greedy algorithm can be implemented using the Hungarian algorithm in O(n^3)

To resolve this issue, all sub-issues must be completed. Before starting work on it, please describe the architectural solution and discuss it with me.

The result of this task should be:
- Three cluster matching strategies implemented: Greedy, Hugarian, and brute force.
- The ability to select a score function for cluster matching strategies.

### LINKED ISSUE #82: FEAT: add Hungarian in cluster matching strategy
This strategy is a more accurate version of the Greedy strategy, but runs in O(n^3) (instead of O(n^2))

### LINKED ISSUE #83: REFACTOR: change API in Initializers module
Remove duplicate code from cluster matching strategies by moving score functions (AIC, BIC, Likelihood etc.) into separate entities
Thus, strategies must take a score function as arguments.

### LINKED ISSUE #84: FEAT: rework the brute force strategy into cluster matching strategies
It's necessary to start considering components from similar families. In other words, iterate over combinations with repetitions.

### PR DESCRIPTION
**Pull request by VITYANA**
# Refactoring of the Initializers Module

## 1. Problem Statement

*   **Code Duplication:** Redundant logic existed across different matching methods.
*   **Inflexibility:** The architecture did not support the combination of arbitrary scoring functions (e.g., AIC, Log-Likelihood) with different cluster-to-model matching strategies.
*   **Missing Functionality:** There was no unified interface for extending functionality, and optimal matching algorithms, such as the Hungarian algorithm, were absent.

## 2. Objectives
The primary goal of this refactoring was to redesign the `initializers` module to achieve the following:

*   **Architectural Decoupling:** Separate the logic for matching clusters to models from the logic used to score those matches.
*   **Unified Interface:** Create a standardized interface allowing users to configure the initialization process.
*   **Algorithm Implementation:** Implement and integrate the Hungarian algorithm and a Permutation-based.

### Architecture
The module has been reorganized into a new `clusterize` subpackage to enforce separation of concerns. The monolithic logic has been decomposed into three independent components:

1.  **Matching Algorithms (`_cluster_match_algorithms.py`):** Encapsulates the logic for assigning data clusters to distribution models.
2.  **Scoring Functions (`_score_functions.py`):** Contains independent implementations of statistical metrics.
3.  **Estimation Strategies (`_estimation_strategies/`):** Defines methods for estimating distribution parameters (e.g., Q-function optimization).
4.  **Utils (`utils.py`):** Contains different useful functions and typings.

### Algorithms and Methods

**Matching Strategies (`MatchingMethod`)**
*   **Greedy Algorithm:** Iteratively assigns the best available cluster to the current model. This method was refactored for improved integration.
*   **Hungarian Algorithm:** Implemented using `scipy.optimize.linear_sum_assignment`. This solves the linear assignment problem by minimizing the total cost (negative score) of individual model-cluster pairs, providing an optimal solution for the sum of individual scores.
*   **Permutations:** A brute-force strategy that generates all possible assignment permutations to evaluate the global score of the entire `MixtureModel`. This guarantees a global optimum but is computationally constrained to a small number of components.

**Scoring Functions (`ScoringMethod`)**
Independent functions were implemented to evaluate goodness-of-fit, usable by any matching strategy:
*   **Log-Likelihood:** Measures the probability of the parameters given the data.
*   **AIC (Akaike Information Criterion):** Penalizes model complexity to prevent overfitting.

## 4. Testing and Quality Assurance
Given the scope of the refactoring, the unit test suite was rewritten and significantly expanded.

*   **Coverage:** Code coverage reached **99%**, exceeding the library's minimum requirement of 80%.
*   **Test Scenarios:**
    *   Verification of the Hungarian and Permutations algorithms against known baselines.
    *   Comparative analysis of different matching algorithms on identical datasets.
    *   **Edge Case Handling:** Validation of system behavior with empty clusters, mismatches between the number of clusters and mixture components, and metric calculations in degenerate cases.
    *   Default workflow scenarios check.

## 5. Related Issues
This work addresses the following project issues:

*   #81: Architecture design and refactoring.
*   #82: Implementation of the Hungarian algorithm.
*   #83: Implementation of independent AIC and Log-Likelihood scoring functions.
*   #84: Implementation of the Permutations matching algorithm.

## 6. P.S
All functions, methods, modules and classes provided with documentation, so you can check it in detail

### CODE REVIEW

File: rework_pysatl_mpest/initializers/__init__.py
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/__init__.py:1]: In examples we use imports from modules, not files inside module. That's why there is `__init__.py` files

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 1]
```diff
@@ -16,6 +16,43 @@
 from rework_pysatl_mpest.optimizers import Optimizer, ScipyNelderMead
 
 
+def _validate_clusters_distributions(
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]: Missed docstrings

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 2]
```diff
@@ -16,6 +16,43 @@
...
+
+    cluster_weights: list[float] = np.sum(H, axis=0)
+
+    valid_clusters = [k for k in range(n_clusters) if cluster_weights[k] >= min_samples]
+    if len(valid_clusters) != models_count:
+        return [], []
+    return valid_clusters, cluster_weights
+
+
+def _calculate_cluster_fit(
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]: Missed docstrings

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 3]
```diff
@@ -16,6 +16,43 @@
...
+    temp_model: ContinuousDistribution,
+    estimation_func: Callable,
+    X: np.ndarray,
+    H_k: np.ndarray,
+    optimizer: Optimizer,
+) -> tuple[dict[str, float], float]:
+    new_params = estimation_func(temp_model, X, H_k, optimizer)
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]:
```suggestion
    param_names, param_values = estimation_func(temp_model, X, H_k, optimizer).items()
```

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 4]
```diff
@@ -16,6 +16,43 @@
...
+    param_names = new_params.keys()
+    param_values = new_params.values()
+    temp_model.set_params_from_vector(param_names, param_values)
+
+    log_probs = np.clip(temp_model.lpdf(X), -1e9, -1e-9)
+    weighted_log_likelihood = np.sum(H_k * log_probs)
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]:
```suggestion
    weighted_log_likelihood = np.dot(H_k, log_probs)
```

File: rework_pysatl_mpest/initializers/clusterize_initializer.py [hunk 5]
```diff
@@ -320,6 +320,8 @@ def perform(
         5. Returns the initialized mixture model
         """
         X = np.asarray(X, dtype=np.float64)
+        if X.ndim == 1:
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/clusterize_initializer.py]: Move this inside `self._clusterize(...)`
Otherwise, your two-dimensional array X will be passed to the rest of the class's internal methods, which seem not to be suited for this.

File: rework_tests/unit/test_initializers/test_clusterize_initializer.py [hunk 6]
```diff
@@ -183,7 +183,7 @@ def test_perform_accurate_init_normal_path(self):
     def test_perform_accurate_init_fallback_to_fast_init(self):
         initializer = ClusterizeInitializer(is_accurate=True, is_soft=True, clusterizer=self.mock_clusterizer)
 
-        X = np.array([1.0, 2.0, 3.0])
+        X = np.array([[1.0], [2.0], [3.0]])
```
Comment (iraedeus) [inline rework_tests/unit/test_initializers/test_clusterize_initializer.py]: For what?
This test is not intended to check various X formats.

Besides, why this particular test? It seems to me that tests should be written in the format the user will work with, assuming a one-dimensional array will be passed.

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 7]
```diff
@@ -77,52 +114,32 @@ def match_clusters_for_models_log_likelihood(
...
-
-            log_probs = np.clip(temp_model.lpdf(X), -1e9, -1e-9)
-            weighted_log_likelihood = np.sum(H_k * log_probs)
+            H_k = H[:, k]
+            new_params, weighted_log_likelihood = _calculate_cluster_fit(
+                copy(model), estimation_func, X, H_k, optimizer
+            )
 
             effective_n = cluster_weights[k]
             score = weighted_log_likelihood / effective_n
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]: Why is there an additional division here?

File: rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py:1]: Refactor the constructor and the `perform` signature:
The user should pass the default optimizer and clusterer to the constructor, and pass them to `perform` if they want to use other ones.

Comment (iraedeus) [reply rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py:1]: It is necessary to add a check that the clusterer has the required methods in the constructor and in `perform`, since the object type is `Any`

**Review by iraedeus — CHANGES_REQUESTED** [review]

File: rework_pysatl_mpest/initializers/initializer.py [hunk 8]
```diff
@@ -52,7 +52,8 @@ def perform(
         self,
         X: ArrayLike,
         dists: list[ContinuousDistribution],
-        cluster_match_info: ClusterMatchStrategy,
+        cluster_match_info: MatchingMethod,
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/initializer.py]: forgot to update the type hint in the docs

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 9]
```diff
@@ -8,269 +8,299 @@
...
+    cluster_weights = context["cluster_weights"]
+    n_samples = len(X)
+
+    cached_fits: list[list[FitResult]] = []
+    computation_cache: dict[tuple, FitResult] = {}
+
+    for model, est_func in zip(models, estimation_strategies):
+        row: list[FitResult] = []
+        for k in valid_clusters:
+            cache_key = (model.__class__, est_func, k)
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]: Idk, but what happens if models contains multiple instances of the same class? It seems like they would share the same cache key.

File: rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py [hunk 10]
```diff
@@ -285,7 +282,8 @@ def perform(
         self,
         X: ArrayLike,
         dists: list[ContinuousDistribution],
-        cluster_match_strategy: ClusterMatchStrategy,
+        method: MatchingMethod,
+        score_func: ScoringMethod,
         estimation_strategies: list[EstimationStrategy],
         optimizer: Optimizer = ScipyNelderMead(),
     ) -> MixtureModel:
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py:393]: The argument names here do not match the abstract base class definition in Initializer

File: rework_pysatl_mpest/initializers/clusterize/_cluster_match_algorithms.py [hunk 11]
```diff
@@ -8,269 +8,299 @@
...
+    return cached_fits
+
+
+def _match_greedy(context: Context) -> MatchingResult:
+    """sequentially assign each model to its best available cluster"""
+    updated_params_list: list[dict[str, float]] = []
+    model_weights: list[float] = []
+    used_clusters = set()
+
+    for model, estimation_func in zip(context["models"], context["estimation_strategies"]):
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/clusterize/_cluster_match_algorithms.py:66]: The greedy strategy currently iterates sequentially. Was this order-dependency intended?

File: rework_pysatl_mpest/initializers/cluster_match_strategy.py [hunk 12]
```diff
@@ -8,269 +8,299 @@
...
-            if k in used_clusters:
-                continue
-            H_k = H[:, k]
+    score = score_func(temp_model, X, H_k)
 
-            new_params = estimation_func(temp_model, X, H_k, optimizer)
-            param_names = new_params.keys()
-            param_values = new_params.values()
-            temp_model.set_params_from_vector(param_names, param_values)
+    return {"model": temp_model, "params": new_params, "score": score, "weight": -1.0}
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/cluster_match_strategy.py]: You can estimate weight here by using `H_k`

File: rework_pysatl_mpest/initializers/clusterize_initializer.py [hunk 13]
```diff
@@ -89,7 +89,7 @@ class ClusterizeInitializer(Initializer):
         {EstimationStrategy.QFUNCTION: q_function_strategy}
     )
 
-    def __init__(self, is_accurate: bool, is_soft: bool, clusterizer: Any):
+    def __init__(self, is_accurate: bool, is_soft: bool, clusterizer: Any = None, optimizer: Optimizer | None = None):
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/clusterize_initializer.py]: Why clusterizer and optimizer here is optional? Require the user to specify these fields in the constructor, and then you won't have to check for their presence in the `perform` method.

File: rework_pysatl_mpest/initializers/initializer.py
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/initializer.py:1]: The abstract initializer class's description is fundamentally incorrect.
You've literally described the clustering logic here, although in fact, the initializer can initialize distributions in its own way, not necessarily through clustering. Therefore, I think the signature of the `perform` method should be as follows:

```python
@abstractmethod
    def perform(
        self,
        X: ArrayLike,
        dists: list[ContinuousDistribution]
    ) -> MixtureModel:
```

Also you forgot to add typing for returned value.

Comment (iraedeus) [reply rework_pysatl_mpest/initializers/initializer.py:1]: So all initializers work with `X` and the `dists` list, but for example ClusterizeInitializer will additionally work with `method`, `score_func` and `estimation_strategies`

File: rework_pysatl_mpest/initializers/clusterize/_cluster_match_algorithms.py
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/clusterize/_cluster_match_algorithms.py:1]: For some reason, not all functions have full documentation (some are missing the Parameters and Returns sections)

**Review by iraedeus — CHANGES_REQUESTED** [review]
I'd suggest slightly changing the module's file structure, moving all files related to 'ClusterizeInitializer' to a separate directory. Something like this:

initializers/
├── _estimation_strategies/
├── clusterize/
│   ├── cluster_match_strategy.py
│   ├── clusterize_initializer.py
│   └── strategies.py
├── __init__.py
└── initializer.py

I also still don't like that everything is mixed together in one file, `cluster_match_strategy.py`. I would split it into three files: `cluster_match_algorithms.py` (contains three algorithms: Greedy, Hungarian, and brute force), `score_functions.py` (contains AIC, BIC, and Likelihood), and `utils.py` or something like that for helper functions (I need to think about this).

File: rework_pysatl_mpest/initializers/__init__.py [hunk 14]
```diff
@@ -16,9 +16,9 @@
 __license__ = "SPDX-License-Identifier: MIT"
 
 from ._estimation_strategies.q_function import q_function_strategy, q_function_strategy_exponential
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/__init__.py]: Remove this from `__all__`, module `_strategies` is internal

File: rework_pysatl_mpest/initializers/clusterize/score_functions.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+def _calculate_component_log_likelihood(model: ContinuousDistribution, X: np.ndarray, H_k: np.ndarray) -> float:
+    """Calculates the weighted log-likelihood for a single component"""
+    log_probs = np.clip(model.lpdf(X), -1e9, -1e-9)
+    return np.sum(H_k * log_probs)
+
+
+def _calculate_mixture_log_likelihood(model: MixtureModel, X: np.ndarray) -> float:
+    """Calculates the total log-likelihood for a mixture model"""
+    return np.sum(model.loglikelihood(X))
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/clusterize/score_functions.py]: `MixtureModel.loglikelihood` returns scalar, so:
```suggestion
    return model.loglikelihood(X)

File: rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py [hunk 16]
```diff
@@ -303,25 +303,23 @@ def perform(
...
-            The clustering algorithm instance (e.g., KMeans, C-Means).
-            If not provided, the default clusterizer passed to __init__ is used.
+        method : MatchingMethod, optional
+            The algorithm used to match the clusters. Default is GREEDY.
+        score_func : ScoringMethod, optional
+            The metric used to score the fit. Default is LIKELIHOOD.
+        estimation_strategies : list[EstimationStrategy], optional
+            List of estimation strategies. If None, uses QFUNCTION for all models.
+        optimizer : Optimizer, optional
+            Optimizer for parameter estimation.
```
Comment (iraedeus) [inline rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py]:
```suggestion
Optimizer for parameter estimation. If None, uses the one from __init__.
```

**Review by iraedeus — CHANGES_REQUESTED** [review]
I think it is better to move `match_clusters_for_models` inside `ClusterizeInitializer` as well as the dictionaries `_MATCHING_METHOD`, `_SCORING_METHOD`.

I also found a bug: you're using a Q-function with a minus sign as the component's score function, which is fundamentally wrong, since a larger function value should indicate a better approximation. The larger the Q-function, the better the component fits the cluster, while a minus sign does the opposite. Your score functions are described by the logic "maximum is better."

So `cluster_match_algorithms.py` and `score_functions.py` became private modules

**Review by iraedeus — APPROVED** [review]

File: rework_pysatl_mpest/initializers/__init__.py [hunk 17]
```diff
@@ -9,64 +9,24 @@
 
 initializers provide good starting points for EM algorithm and other optimization
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/__init__.py]:
```suggestion
Initializers provide good starting points for EM algorithm and other optimization
```

File: rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py [hunk 18]
```diff
@@ -285,9 +384,12 @@ def perform(
...
         dists: list[ContinuousDistribution],
-        cluster_match_strategy: ClusterMatchStrategy,
-        estimation_strategies: list[EstimationStrategy],
-        optimizer: Optimizer = ScipyNelderMead(),
+        method: MatchingMethod = MatchingMethod.GREEDY,
+        score_func: ScoringMethod = ScoringMethod.LIKELIHOOD,
+        estimation_strategies: list[EstimationStrategy] | None = None,
+        optimizer: Optimizer | None = None,
+        clusterizer: Clusterizer | None = None,
+        **kwargs: Any,
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py:392]: I'm not entirely sure, but it seems like `**kwargs` can be deleted here.

File: rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py [hunk 19]
```diff
@@ -167,21 +185,92 @@ def _clusterize(self, X: np.ndarray, clusterizer: Any) -> np.ndarray:
...
+        Notes
+        -----
+        The function follows these steps:
+        1. Calculates minimum sample threshold (1% of data size).
+        2. Validates clusters and filters out those with insufficient samples.
+        3. Prepares the execution context.
+        4. Dispatches to the specific matching algorithm defined by `method`.
+        """
+        n_models = len(models)
+        min_samples = int(np.ceil(len(X) * 0.01))
```
Comment (xImoZA) [inline rework_pysatl_mpest/initializers/clusterize/clusterize_initializer.py:248]: Maybe it's worth moving this threshold to the function parameters or the class constructor.

**Review by xImoZA — APPROVED** [review]
LGTM
